### DISCUSSION TITLE
Hooks inside functions (in other packages)

### DISCUSSION THREAD

**Discussion by fonsp** [JuliaPluto/PlutoHooks.jl#8] [Q&A]
It would be really cool if in this example, both calls to `f` would have their own ref. This would mean that you can "package up" a component written using hooks into a function (in another package), and people just call the function.

https://user-images.githubusercontent.com/6933510/144312629-93281aba-de3c-49ea-b4a9-26a609de2c3d.mov

**Comment by Pangoraw** [comment]
I think this would maybe be tricky to identify the different call sites to associate state with each one unlike with a macro where each call has its own code (expr). I think @dralletje made a `@hook` macro that simplifies the creation of hooks and make them look more functional at [PlutoHooks.jl#@hook-hook](https://github.com/JuliaPluto/PlutoHooks.jl/tree/%40hook-hook).

**Comment by fonsp** [comment]
I'm thinking more about this, there are lots of use cases that only work if PlutoHooks can be used "behind the scenes" in a function call or `Base.show` method inside a package. 

It also makes it easier to write other hooks that use PlutoHooks, right now that's pretty complex with all the escaping.

Is it possible to make all the hooks available as a function? It could be implemented like preact: every cell has a counter of which hook is being called? 

@schlichtanders Already uses PlutoHooks with functions instead of macros, right? How does that work?

**Reply by schlichtanders** [reply]
Hi there,
the key ingredient why I can use functions is that Pluto strictly executes one cell after another. There is not a single moment where two cells are run in parallel. I think this design is very useful also mentally so that users do not get too much confused about the semantics of code execution. 

Given this restriction, one can use global state to empower the hooks.

There are already existing `Ref`s, notably the one which contains the currently running cell id.
In addition I added another `Ref` which captures whether the current cell execution is triggered by human or by other cells or reactivity (this is already an existing variable, but it was not needed as a state because the macro execution made this redundant - the macro was always only expanded on human triggered execution).

With these I made all PlutoHooks work from Python which does not support macros. (That was my initial motivation).

**Reply by schlichtanders** [reply]
Here my own core PlutoHook implementation, which can give you an idea about how it all works together

https://github.com/jolin-io/JolinPluto.jl/blob/main/src/tasks.jl#L148-L211
